fix(curriculum): enhanced grammer and flexibility in tests to accommodate any valid response (#53786)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/653639d63a45a077333312c8.md

# --description--

Before playing the song, you need to make sure it starts from the beginning. This can be achieved by the use of the `currentTime` property on the `audio` object. 

> Add an `if` statement to check whether the `userData?.currentSong` is falsy *OR* if `userData?.currentSong.id` is strictly not equal to `song.id`. This condition will check if no current song is playing or if the current song is different from the one that is about to be played.

Reviewer: ```suggestion
You should create an `if` statement that checks whether `userData.currentSong` is either falsy *OR* `userData.currentSong` is not equal to `song.id`. Don't forget to use optional chaining.
```

Inside your `if` block, set the `audio.currentTime` property to `0`.

# --hints--

> You need to construct an `if` statement that verifies whether `userData?.currentSong` is falsy (either null or undefined) and also confirms that `userData?.currentSong.id` doesn't match the `id` of the song.

Author: @Ksound22 Done.

# --description--

Before playing the song, you need to make sure it starts from the beginning. This can be achieved by the use of the `currentTime` property on the `audio` object. 

Add an `if` statement to check whether the `userData?.currentSong` is falsy *OR* if `userData?.currentSong.id` is strictly not equal to `song.id`. This condition will check if no current song is playing or if the current song is different from the one that is about to be played.

Inside your `if` block, set the `audio.currentTime` property to `0`.

# --hints--

> You should create an `if` statement that checks whether `userData.currentSong` is either falsy *OR* `userData.currentSong` is not equal to `song.id`. Don't forget to use optional chaining.

Reviewer: Sorry I didn't catch this on time

```suggestion
You should create an `if` statement that checks whether `userData.currentSong` is either falsy *OR* `userData.currentSong.id` is not equal to `song.id`. Don't forget to use optional chaining.
```

```js

> assert.match(code, /if\s*\(\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\)\s*\{\s*/);

Reviewer: Since the first test checked for the condition and the various ways it can come in, I don't think we need to bring that in here again.

This looks tricky but it gets the job done. It checks for `audio.currentTime = 0;` inside curly braces.  

```suggestion
assert.match(code, /\{\s*audio\.currentTime\s*=\s*0;?\s*\}/);
```

Reviewer 2: As much as this is shorter, unfortunately it doesn't check where exactly it was added.

Author: @Ksound22 We first need to account for all the possible code variations for `if` condition, only then we can check for `audio.currentTime = 0` inside of curly braces.

> assert.match(code, /if\s*\(\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\)\s*\{\s*/);

Reviewer 2: ```suggestion
assert.match(code, /if\s*\(\s*(?:(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined))\s*\)\s*\{/);
```
There's missing one grouping, what allows passing with only part of code:
```javascript
if (!userData?.currentSong || userData?.currentSong.id !== song.id
```
```javascript
userData?.currentSong.id !== song.id || !userData?.currentSong ) {
```

Author: @gikf The code snippets you've provided work even without this suggestion. Please explain what this suggestion does.

Reviewer 2: @aaqib605 That's the issue, they shouldn't pass, as that's invalid code.

Author: @gikf Got it. I made the necessary changes, and now it should work fine.

Reviewer 2: Are you sure? It doesn't look like anything was changed...

Author: @gikf Could you provide a screenshot highlighting the issue with the code?

Reviewer 2: Line 21:
https://github.com/freeCodeCamp/freeCodeCamp/blob/7d9c3be23793b7ce5090649961a967ed2daf33b4/curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/653639d63a45a077333312c8.md?plain=1#L21
As in the first comment with suggestion how to change it below:
```suggestion
assert.match(code, /if\s*\(\s*(?:(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined))\s*\)\s*\{/);
```

- - -
Similarly with https://github.com/freeCodeCamp/freeCodeCamp/pull/53786/files#r1499699708, no changes were made, semicolon is still required by regex.
https://github.com/freeCodeCamp/freeCodeCamp/blob/7d9c3be23793b7ce5090649961a967ed2daf33b4/curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/653639d63a45a077333312c8.md?plain=1#L29

```

You should set `audio.currentTime` to `0` inside your `if` block.

```js
assert.match(
  code,

>   /if\s*\(\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\)\s*\{\s*audio\.currentTime\s*=\s*0;\s*\}/

Reviewer 2: ```suggestion
  /if\s*\(\s*(?:(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined))\s*\)\s*\{\s*audio\.currentTime\s*=\s*0\s*;?\}/
```
Corrected the part of regex, plus optional semicolon.

Author: @gikf Implemented this. Thanks for the suggestion.

# --description--

Before playing the song, you need to make sure it starts from the beginning. This can be achieved by the use of the `currentTime` property on the `audio` object. 

Add an `if` statement to check whether the `userData?.currentSong` is falsy *OR* if `userData?.currentSong.id` is strictly not equal to `song.id`. This condition will check if no current song is playing or if the current song is different from the one that is about to be played.

Inside your `if` block, set the `audio.currentTime` property to `0`.

# --hints--

> You should create an `if` statement that checks whether `userData.currentSong` is either falsy *OR* `userData.currentSong.id` is not equal to `song.id`. Don't forget to use optional chaining.

Reviewer 3: ```suggestion
You should create an `if` statement that checks whether `userData?.currentSong` is either falsy *OR* `userData?.currentSong.id` is not equal to `song.id`. Don't forget to use optional chaining.
```

Reviewer 4: uhm, no, this still passes without the required assignment:
![image](https://github.com/freeCodeCamp/freeCodeCamp/assets/26656284/d6554252-275d-47af-98a6-cad5a597473b)

Reviewer 4: oh, wait, I had an outdated version locally. Sorry I didn't want to send that commit, let me fix that

```js

> assert.match(code, /if\s*\(\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\)\s*\{\s*/);

Reviewer 3: I've seen the issue and read the conversations, I've come up with a solution that satisfies all condiitons,

```suggestion
assert.match(code, /if\s*\(\s*(?:userData\?\.currentSong\s*===\s*(?:null|undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id|userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*userData\?\.currentSong\s*===\s*(?:null|undefined))\s*\)\s*\{\s*/);
```

look here, there's 3 blocks mainly 
first condition
```
userData\?\.currentSong\s*===\s*(?:null|undefined)
```
the or statement
```
\s*\|\|\s*
```
last condition
```
 userData\?\.currentSong\.id\s*!==\s*song\.id
```

I've written and simplified the regex in a way that it targets both cases of "Either this or that". try this solution @aaqib605

> assert.match(code, /if\s*\(\s*(?:(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined))\s*\)\s*\{/);

Reviewer 4: ```suggestion
assert.match(code, /if\s*\(\s*(?:(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*undefined))\s*\)\s*\{/);
```

```

You should set `audio.currentTime` to `0` inside your `if` block.

```js

> assert.match(
>   code,
>   /if\s*\(\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\)\s*\{\s*audio\.currentTime\s*=\s*0;\s*\}/

Reviewer 3: applying the same theory here as well 

```suggestion
assert.match(
  code,
  /if\s*\(\s*(?:userData\?\.currentSong\s*===\s*(?:null|undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id|userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*userData\?\.currentSong\s*===\s*(?:null|undefined))\s*\)\s*\{\s*audio\.currentTime\s*=\s*0;?\s*\}/
);
```

>    /if\s*\(\s*(?:(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined))\s*\)\s*\{\s*audio\.currentTime\s*=\s*0\s*;?\s*\}/

Reviewer 4: ```suggestion
   /if\s*\(\s*(?:(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*undefined))\s*\)\s*\{\s*audio\.currentTime\s*=\s*0\s*;?\s*\}/
```
